Fix models_config.json file issue

S1LV3RJ1NX opened the pull request:
Ignore registering models if API key is not present, and put out a war…ning. Also create a config.json file at runtime, helpful for template deployments

chiragjn commented on backend/modules/model_gateway/__init__.py:
```diff
@@ -1 +1,19 @@
+
+# Check if models_config.sample.json exists but not models_config.json
+# If models_config.sample.json exists, copy it to models_config.json
+if os.path.exists("./models_config.sample.json") and not os.path.exists(
+    "./models_config.json"
+):
```
This can break depending on what is the cwd, the path should be made relative to this file using os.path operations on `__file__`

S1LV3RJ1NX replied:
changed.

chiragjn commented on backend/modules/model_gateway/model_gateway.py:
```diff
@@ -1,85 +1,119 @@
-    def __init__(self):        
+
+    def __init__(self):
         with open("./models_config.json") as f:
```
Same comment about paths here

S1LV3RJ1NX replied:
done.

chiragjn commented on backend/modules/model_gateway/model_gateway.py:
```diff
@@ -1,85 +1,119 @@
             data = json.load(f)
             self.config = [ModelProviderConfig(**item) for item in data]
             self.models: List[ModelConfig] = []
             for list in self.config:
```
We should not be using `list` as variable name as it shadows built in type name

S1LV3RJ1NX replied:
done.

chiragjn commented on backend/modules/model_gateway/model_gateway.py:
```diff
@@ -1,85 +1,119 @@
+    def __init__(self):
         with open("./models_config.json") as f:
             data = json.load(f)
             self.config = [ModelProviderConfig(**item) for item in data]
```
nit: this should have been `.parse_obj(item)`

S1LV3RJ1NX replied:
done.

chiragjn commented on backend/modules/model_gateway/model_gateway.py:
```diff
@@ -1,85 +1,119 @@
+                    else:
+                        # Avoid adding the model to the map if the API key is not set
+                        logger.warning(
+                            f"Environment variable {list.api_key_env_var} not set."
```
This error message should be more helpful
We should inform we are skipping adding the model to list of models

S1LV3RJ1NX replied:
done.

chiragjn commented on backend/modules/model_gateway/model_gateway.py:
```diff
@@ -1,85 +1,119 @@
+                    else:
+                        # Avoid adding the model to the map if the API key is not set
+                        logger.warning(
+                            f"Environment variable {list.api_key_env_var} not set."
```
Same as above

S1LV3RJ1NX replied:
done.

chiragjn commented on backend/modules/model_gateway/model_gateway.py:
```diff
@@ -1,85 +1,119 @@
 
 class ModelGateway:
     config: List[ModelProviderConfig]
     modelsToProviderMap = {}
```
nit: convert to snake case
naming is also pretty confusing, isn't this `model_id_to_provider_config` ?

S1LV3RJ1NX replied:
@innoavator: Please check on this.

chiragjn commented on backend/modules/model_gateway/model_gateway.py:
```diff
@@ -1,85 +1,119 @@
+
     def get_embedding_models(self) -> List[ModelConfig]:
         models: List[ModelConfig] = []
         for list in self.config:
```
Why not compute this just once in `__init__`?

S1LV3RJ1NX replied:
Done.

chiragjn commented on backend/modules/model_gateway/model_gateway.py:
```diff
@@ -1,85 +1,119 @@
+
 
 class ModelGateway:
     config: List[ModelProviderConfig]
     modelsToProviderMap = {}
```
Do we get any benefit from making these static?

S1LV3RJ1NX replied:
@innoavator: Please check on this.

chiragjn commented on backend/modules/model_gateway/model_gateway.py:
```diff
@@ -1,85 +1,119 @@
+                    models.append(
+                        ModelConfig(
+                            name=f"{list.provider_name}/{model_id}", type=ModelType.chat
+                        ).dict()
```
Why `dict` ?
the return type says `List[ModelConfig]`

S1LV3RJ1NX replied:
it was leading to JSON serialization error. Added a `to_dict()` to pydantic

chiragjn commented on backend/modules/model_gateway/model_gateway.py:
```diff
@@ -1,85 +1,119 @@
+                        ModelConfig(
+                            name=f"{list.provider_name}/{model_id}",
+                            type=ModelType.embedding,
+                        ).dict()
```
Same here, why `.dict()` ?

chiragjn commented on backend/modules/model_gateway/model_gateway.py:
```diff
@@ -1,85 +1,119 @@
         else:
-            api_key = os.environ.get(model_provider_config.api_key_env_var, '')
+            api_key = os.environ.get(model_provider_config.api_key_env_var, "")
         model_id = "/".join(model_name.split("/")[1:])
```
We should be clear what is considered id and what is considered name
just from reading I would assume id is the fully qualified name but here is seems reverse

chiragjn commented on backend/modules/model_gateway/model_gateway.py:
```diff
@@ -1,85 +1,119 @@
         else:
-            api_key = os.environ.get(model_provider_config.api_key_env_var, '')
+            api_key = os.environ.get(model_provider_config.api_key_env_var, "")
         model_id = "/".join(model_name.split("/")[1:])
```
```suggestion
        _, model_id = model_name.split("/", 1)
```

S1LV3RJ1NX replied:
@innoavator: Please check this

innoavator reviewed (changes requested):
We cannot merge this PR - we are just delaying the failure instead of asking the user to fix the issue. We should make the code independent of the model and power all the models via the gateway.

S1LV3RJ1NX commented:
> We cannot merge this PR - we are just delaying the failure instead of asking the user to fix the issue. We should make the code independent of the model and power all the models via the gateway.

The above changes were mostly done to allow the code to be executed on the platform too! One way I can see is, raise errors if models are not defined, and for TFY deployment purposes, generate the config.json at runtime. Does that work?

S1LV3RJ1NX commented:
> We cannot merge this PR - we are just delaying the failure instead of asking the user to fix the issue. We should make the code independent of the model and power all the models via the gateway.

@innoavator  I have now changed the config. User cannot proceed without setting up the config file correctly. Also fixed the config file generation issue while deployment. Please check some unresolved comments wrt to variable naming.

chiragjn commented on backend/types.py:
```diff
@@ -104,6 +104,13 @@ class ModelConfig(BaseModel):
     type: Optional[ModelType]
     parameters: Optional[Dict[str, Any]] = None
 
+    def to_dict(self) -> Dict[str, Any]:
```
This doesn't seem to be used anywhere?

chiragjn commented on backend/server/routers/internal.py:
```diff
@@ -113,6 +113,8 @@ def get_enabled_models(
 
+    # Serialized models
+    serialized_models = [model.dict() for model in enabled_models]
     return JSONResponse(
-        content={"models": enabled_models},
+        content={"models": serialized_models},
```
This is fine for now, but normally we should define DTO Pydantic objects and just return the objects - fastapi will take care of serializing
